### ISSUE TITLE
nunit3-console crashes at end of multi-fixture test run

### ISSUE THREAD

**Issue by ccarmannt** [nunit/nunit#2840]
When a multi-fixture test is run using nunit3-console via the following command, the tests are all executed, but the nunit3-console executable crashes at the end and fails to write out the results of the run.

This does NOT occur when any individual test fixture is run, but ONLY when running them all.

````
nunit3-console.exe mydll.dll --where=cat==myTestCategory --teamcity --workers=6 --result=myResults.xml
````

Error output:

````
Errors, Failures and Warnings

1) Error :
NUnit.Engine.NUnitEngineException : The NUnit 3 driver encountered an error while executing reflected code.
  ----> System.NullReferenceException : Object reference not set to an instance of an object.
--NUnitEngineException

Server stack trace:
   at NUnit.Engine.Drivers.NUnit3FrameworkDriver.CreateObject(String typeName, Object[] args)
   at NUnit.Engine.Drivers.NUnit3FrameworkDriver.Run(ITestEventListener listener, String filter)
   at NUnit.Engine.Runners.DirectTestRunner.RunTests(ITestEventListener listener, TestFilter filter)
   at System.Runtime.Remoting.Messaging.StackBuilderSink._PrivateProcessMessage(IntPtr md, Object[] args, Object server, Object[]& outArgs)
   at System.Runtime.Remoting.Messaging.StackBuilderSink.SyncProcessMessage(IMessage msg)

Exception rethrown at [0]:
   at System.Runtime.Remoting.Proxies.RealProxy.HandleReturnMessage(IMessage reqMsg, IMessage retMsg)
   at System.Runtime.Remoting.Proxies.RealProxy.PrivateInvoke(MessageData& msgData, Int32 type)
   at NUnit.Engine.ITestEngineRunner.Run(ITestEventListener listener, TestFilter filter)
   at NUnit.Engine.Runners.ProcessRunner.RunTests(ITestEventListener listener, TestFilter filter)
--NullReferenceException
   at NUnit.Framework.Internal.PropertyBag.AddToXml(TNode parentNode, Boolean recursive)
   at NUnit.Framework.Internal.Test.PopulateTestNode(TNode thisNode, Boolean recursive)
   at NUnit.Framework.Internal.TestMethod.AddToXml(TNode parentNode, Boolean recursive)
   at NUnit.Framework.Internal.TestResult.AddToXml(TNode parentNode, Boolean recursive)
   at NUnit.Framework.Internal.TestResult.AddToXml(TNode parentNode, Boolean recursive)
   at NUnit.Framework.Internal.TestResult.AddToXml(TNode parentNode, Boolean recursive)
   at NUnit.Framework.Internal.TestResult.AddToXml(TNode parentNode, Boolean recursive)
   at NUnit.Framework.Internal.TestResult.AddToXml(TNode parentNode, Boolean recursive)
   at NUnit.Framework.Api.FrameworkController.RunTests(ICallbackEventHandler handler, String filter)

Test Run Summary
  Overall result: Failed
  Test Count: 0, Passed: 0, Failed: 0, Warnings: 0, Inconclusive: 0, Skipped: 0
  Start time: 2018-04-30 19:31:36Z
    End time: 2018-04-30 21:07:57Z
    Duration: 5780.826 seconds
````

Framework versions:

NUnit: 3.7.1
NUnit.ConsoleRunner: 3.7.0/3.8.0
NUnit3TestAdapter: 3.6.0
Selenium.WebDriver.IEDriver: 3.8.0

**Comment by ChrisMaddock** [comment]
Would you try upgrading to the latest version of the NUnit framework - and see if you still get the error? This looks like one we’ve seen before.

**Comment by ccarmannt** [comment]
We're looking into doing that, but doing so would require upgrading the server side also, which would impact many teams. So not sure how long it may be before we can complete that effort.

**Comment by ChrisMaddock** [comment]
Ok. In that case - let’s close this until we’ve got a confirmed and reproducible issue for someone to look into. I suggest you try updating, and we can reopen this if you still see the issue. 😊

**Comment by ChrisMaddock** [comment]
Hi @ccarmannt,

You've told us that your test suite throws an exception. Unfortunately - my test suites don't throw this exception! 😄 That makes it very hard for me to guess what you're doing in your code that I'm not doing in mine. You'll need to provide a minimal reproduction of this issue before anyone can help you track down the cause.

Also, we generally don't fix bugs in older versions of NUnit 3.x. As before - I'd advise updating to the latest version of the NUnit framework, and see if your problem persists. If so, I'd then recommend breaking out a minimal example of code that your able to share, and, once we can see the problem, someone may be able to help you work out how to fix it.

Please remember that NUnit is open source software entirely maintained by volunteers. The more you can break down your issue in a way where the issue can be pinpointed and reproduced, the more likely someone is to pick up the issue and take a look at it for you. You're also welcome to debug yourself, and suggest a PR to fix the issue. 

If you choose that course, I'd recommend posting an issue about your bug and proposed fix first, and discussing with the NUnit team, to make sure the fix fits the intended architecture.

**Comment by mikkelbu** [comment]
@ccarmannt The best match I could find in the existing issues is https://github.com/nunit/nunit/issues/1958. Do you have any place where you add anything to the `PropertyBag` similar to the example in https://github.com/nunit/nunit/issues/1958#issuecomment-270657221?

Do you get any information by running the tests `inprocess` or by setting the `trace`-level to something higher, like `Verbose`.

**Comment by ccarmannt** [comment]
@mikkelbu when I run with trace set to Verbose, I see a couple of errors in the log like this one:

````
15:52:44.187 Error [ 8] TestProgressReporter: Exception processing MyProject.FeatureFiles.MenuNavigationExternalUserTests.ExternalUser
System.NullReferenceException: Object reference not set to an instance of an object.
   at NUnit.Framework.Internal.PropertyBag.AddToXml(TNode parentNode, Boolean recursive)
   at NUnit.Framework.Internal.Test.PopulateTestNode(TNode thisNode, Boolean recursive)
   at NUnit.Framework.Internal.TestMethod.AddToXml(TNode parentNode, Boolean recursive)
   at NUnit.Framework.Internal.TestResult.AddToXml(TNode parentNode, Boolean recursive)
   at NUnit.Framework.Internal.TestProgressReporter.TestFinished(ITestResult result)
````

We do have some tests that add values to the PropertyBag as noted in the #1958 comment. However, since the crashing behavior does not occur when any single fixture is run, I would think this is a different problem.

**Comment by CharliePoole** [comment]
You can tie this to individual tests by examining the log. Each entry shows the thread on which the event occurred. Track backwards in the log to see what test just finished on that thread. Note that "test" may mean an individual case or a suite here.

**Comment by ccarmannt** [comment]
I was able to confirm via a reduced scenario that this can happen when running a single fixture.

Is it a good idea, as proposed elsewhere, to use `TestExecutionContext.CurrentContext.CurrentTest.Properties.Add()` instead of `TestContext.CurrentContext.Test.Properties.Add()`?  Or is that likely to be disabled at a future point as well?

I suppose this should be closed as a duplicate of #1958.

**Comment by rprouse** [comment]
I am closing old uncategorized issues that have not had comments or made progress in several years. If anyone comes back with a compelling argument for these issues, we can reopen.
